File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_measurement_unit_mappers/EnergyMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_measurement_unit_mappers

import androidx.health.connect.client.units.Energy
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.EnergyDto
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

/**
 * Unit tests for Energy Mapper.
 *
 * Tests verify proper bidirectional mapping between [EnergyDto] and Health Connect
 * [Energy] objects, including unit conversions between kilocalories, kilojoules, calories, and joules.
 */
@DisplayName("EnergyMapper")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class EnergyMapperTest {

    private companion object {
        const val TEST_VALUE = 100.0
    }

    @Test
    @DisplayName(
        "GIVEN EnergyDto in kilocalories → " +
            "WHEN toHealthConnect called → " +
            "THEN creates Energy with correct value",
    )
    fun whenEnergyDtoInKilocalories_thenCreatesCorrectEnergy() {
        // Given
        val dto = EnergyDto(kilocalories = TEST_VALUE)

        // When
        val result = dto.toHealthConnect()

        // Then
        result.inKilocalories shouldBe TEST_VALUE
    }

    @Test
    @DisplayName(
        "GIVEN Energy object → " +
            "WHEN toDto called → " +
            "THEN converts to DTO in kilocalories",
    )
    fun whenEnergyToDto_thenUsesKilocalories() {
        // Given
        val energy = Energy.kilocalories(TEST_VALUE)

        // When
        val result = energy.toDto()

        // Then
        result.kilocalories shouldBe TEST_VALUE
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/measurement_unit_mappers/energy_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/energy_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

void main() {
  group(
    'EnergyMapper',
    () {
      group(
        'EnergyToDto',
        () {
          test(
            'converts Energy to EnergyDto in kilocalories',
            () {
              const energy = Energy.kilocalories(500.0);
              final dto = energy.toDto();

              expect(dto.kilocalories, 500.0);
            },
          );

          test(
            'converts Energy from kilojoules to kilocalories',
            () {
              const energy = Energy.kilojoules(2092.0);
              final dto = energy.toDto();

              expect(dto.kilocalories, closeTo(500.0, 0.5));
            },
          );
        },
      );

      group(
        'EnergyDtoToDomain',
        () {
          test(
            'maps EnergyDto to Energy',
            () {
              final dto = EnergyDto(kilocalories: 500.0);
              final energy = dto.toDomain();

              expect(energy.inKilocalories, 500.0);
            },
          );
        },
      );
    },
  );
}
